feat - Support webpack 5 emit

bebraw opened the pull request:
Closes #40.

Note that I'm testing only against webpack 5 at the moment. I can add webpack 4 tests as well though that will complicate the testing setup a bit (we'll need to differentiate between 4 and 5 in directory structure). If webpack 5 only is enough, then the current solution will do.

When running the tests, you see something like this:

```
(node:13253) [DEP_WEBPACK_JAVASCRIPT_MODULES_PLUGIN] DeprecationWarning: webpack.JavascriptModulesPlugin has moved to webpack.javascript.JavascriptModulesPlugin
(node:13253) [DEP_WEBPACK_LIBRARY_TEMPLATE_PLUGIN] DeprecationWarning: webpack.LibraryTemplatePlugin is deprecated and has been replaced by compilation.outputOptions.library or compilation.addEntry + passing a library option
(node:13253) [DEP_WEBPACK_SINGLE_ENTRY_PLUGIN] DeprecationWarning: SingleEntryPlugin was renamed to EntryPlugin
(node:13253) [DEP_WEBPACK_OPTIONS_DEFAULTER] DeprecationWarning: webpack.WebpackOptionsDefaulter is deprecated and has been replaced by webpack.config.getNormalizedWebpackOptions and webpack.config.applyWebpackOptionsDefaults
```

I'll try to dig where those are coming from

bebraw commented:
Here's the full trace:

```
(node:13894) [DEP_WEBPACK_JAVASCRIPT_MODULES_PLUGIN] DeprecationWarning: webpack.JavascriptModulesPlugin has moved to webpack.javascript.JavascriptModulesPlugin
    at get JavascriptModulesPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:181:4)
    at Function.JavascriptModulesPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_LIBRARY_TEMPLATE_PLUGIN] DeprecationWarning: webpack.LibraryTemplatePlugin is deprecated and has been replaced by compilation.outputOptions.library or compilation.addEntry + passing a library option
    at get LibraryTemplatePlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:191:4)
    at Function.LibraryTemplatePlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_SINGLE_ENTRY_PLUGIN] DeprecationWarning: SingleEntryPlugin was renamed to EntryPlugin
    at get SingleEntryPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:246:4)
    at Function.SingleEntryPlugin (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
(node:13894) [DEP_WEBPACK_OPTIONS_DEFAULTER] DeprecationWarning: webpack.WebpackOptionsDefaulter is deprecated and has been replaced by webpack.config.getNormalizedWebpackOptions and webpack.config.applyWebpackOptionsDefaults
    at get WebpackOptionsDefaulter (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/index.js:271:4)
    at Function.WebpackOptionsDefaulter (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/webpack/lib/util/memorize.js:22:13)
    at Object.<anonymous>.__importStar (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:19:96)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/src/index.ts:2:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
    at Runtime._loadModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:844:12)
    at Runtime.requireModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:694:10)
    at Runtime.requireModuleOrMock (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:865:21)
    at Object.<anonymous> (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/test/plugin.test.ts:4:1)
    at Runtime._execModule (/Users/juhovepsalainen/Projects/github/mini-html-webpack-plugin/node_modules/jest-runtime/build/index.js:1245:24)
```

I wonder if that means we'll have to alter the entry handling as well.

bebraw commented:
Based on the docs

```
Compilation.entries removed
MIGRATION: Use Compilation.entryDependencies instead
```

Interestingly enough there's no mention of `entryDependencies` in webpack source and the type still has `entries` in it.

I can see `asyncEntrypoints` but that's likely for the entries that return a `Promise`. As we don't support that case yet, we probably don't have to care about this field for now.

sapegin commented on src/index.ts:
```diff
@@ -158,38 +162,79 @@ function defaultTemplate({
 
 		Promise.resolve((template || defaultTemplate)(options)).then((source) => {
-			compilation.assets[filename] = new RawSource(source);
+			// eslint-disable-next-line
```
Let's disable a particular rule here, not all of them.

sapegin commented on src/index.ts:
```diff
@@ -158,38 +162,79 @@ function defaultTemplate({
+			chunks,
+		} = this.options;
+		const files = getFiles(compilation.entrypoints, chunks);
+		const options = Object.assign({}, { publicPath }, context, files);
```
```suggestion
		const options = { publicPath, ...context, ...files};
```

sapegin commented on test/__snapshots__/plugin.test.ts.snap:
```diff
@@ -9,7 +9,7 @@ exports[`additional body 1`] = `
     </head>
     <body>
-      <div>Demo</div><script src=\\"runtime~main.js\\"></script><script src=\\"main.js\\"></script>
+      <div>Demo</div><script src=\\"main.js\\"></script>
```
Are these changes intentional?

bebraw replied:
Yes, they have to do with webpack 5 default behavior as there the runtime is included to the output bundles by default.

sapegin reviewed (commented):
Overall looks good to me! I've left a few minor comments.

bebraw commented:
I addressed the comments. I think I'll add a standalone example so we can see if the deprecation warnings appear there as well.

sapegin reviewed (approved):
🧇

bebraw commented:
I think we still have to solve those deprecation warnings before we can merge but it's starting to get there.

bebraw commented:
I figured it out. The issue was that for whatever reason, **mini-css-webpack-plugin** was pulling webpack's old type definitions to the project. Now the type definitions are included to webpack itself.

To solve, I simplified the tests a bit and dropped **mini-css-webpack-plugin**. Let's run with this.
